adapt to Suse/openSUSE

johanneskastl opened the pull request:
~~based on #3~~
adjusts this to run on SUSE/openSUSE 15.x or higher using the python3-targetcli-fb package. Add necessary things to also run on SLES12 or SLES11.

Commit:
add support for SUSE (SLES) and openSUSE Leap

OndrejHome commented on tasks/main.yml:
```diff
@@ -1,20 +1,57 @@
+  package:
+    name: 'python3-targetcli-fb'
+    state: 'installed'
+    disable_recommends: true
```
does this (`disable_recommends`) attribute work with `package` module (https://docs.ansible.com/ansible/latest/modules/package_module.html)? Looking at documentation I see it only in `zypper` module (https://docs.ansible.com/ansible/latest/modules/zypper_module.html#zypper-module). Additionally in `zypper` module it defaults to `yes`/`true` which would mean that we don't need to specify it.

johanneskastl replied:
You are of course right, sorry. This was a leftover of my testing yesterday.

OndrejHome commented on tasks/main.yml:
```diff
@@ -1,20 +1,57 @@
+  failed_when: false
+  changed_when: false
+  check_mode: false
+
```
same feedback as in PR #3 applies here - I think this is not needed here.

johanneskastl replied:
Sorry, this PR is based on #3 and thus contains the same set of changes. Once #3 is merged, I can rebased this PR.

OndrejHome commented on tasks/main.yml:
```diff
@@ -1,20 +1,57 @@
     port: '3260/tcp'
     permanent: true
     state: 'enabled'
+  when: '"firewalld is not installed" not in check_for_firewalld.stdout and "inactive" not in check_for_running_firewalld.stdout'
```
same feedback as in PR #3 applies here - and this can be adjusted to just check if the `firewalld.service` is running

johanneskastl replied:
Sorry, this PR is based on #3 and thus contains the same set of changes. Once #3 is merged, I can rebased this PR.

OndrejHome reviewed (changes requested):
Hi @johanneskastl ,
Thank you also for this PR that adds support for SUSE/opensuse systems.

I have added some comments in code and I would have question about the structure for this change.

**[?]** Do you think that it would be possible to implement package installation and service enable+start more generically? For inspiration below is example on how to include variables for some specific distribution and later re-use them in tasks.

- distribution specific variable import - https://github.com/OndrejHome/ansible.ha-cluster-pacemaker/blob/master/tasks/main.yml#L8-L10
- package installation via variable (generic) - https://github.com/OndrejHome/ansible.ha-cluster-pacemaker/blob/master/tasks/install_normal.yml#L8-L11
- example of variables files for different distributions:
  - https://github.com/OndrejHome/ansible.ha-cluster-pacemaker/blob/master/vars/el8.yml
  - https://github.com/OndrejHome/ansible.ha-cluster-pacemaker/blob/master/vars/fedora30.yml

Above applies only when we can install packages in generic way and only if we have same parameters for service in both systems, which looks to me that we should have (note the comment about `disable_recommends`)

---

Please let me know if you have any questions and/or comments to my suggestions and request.

johanneskastl commented:
> **[?]** Do you think that it would be possible to implement package installation and service enable+start more generically? For inspiration below is example on how to include variables for some specific distribution and later re-use them in tasks.

I was not sure if you are OK with having a massive change to the code, that will introduce new variable files, and rework existing tasks to use the new variables. So I tried to keep the changes to a minimum.

IMHO it is not a clear YES/NO decision if it is worth the effort for just two tasks.

But as I find a variable-based solution a lot more elegant, I will adjust the PR accordingly while rebasing after #3 has been merged.

Commit:
Add new supported platforms to metadata

johanneskastl commented:
I have converted this to use variables, so it is pretty easy to add e.g. Debian or others (at least for the installation tasks).

Needs to be rebased once #3 is merged. Any changes to #3 will need to be included into this PR before it can be merged.

johanneskastl commented:
Please review, now that #3 has been merged and I rebased the changes onto current master.

OndrejHome commented:
Hi @johanneskastl, thank you for adaptation for changed `master`. 
I will be looking at this over this weekend to give you full feedback, but in the meantime could you please let me know on which systems have you tested this PR? (I mean OS versions like 'opensuse 15.1', 'SLES 15.1', etc.).
I'm asking mainly because in metadata section you have noted that this will work on 'all' opensuse versions which I'm worried what 'all' in this context means (all currently supported versions/all existing versions?)

Commit:
ADD vars_SUSE_SLES12.yml containing the different package name on SLES12

Commit:
ADD package python-dbus-python to vars_SUSE_SLES12.yml on SLES12 systems

Commit:
fix message as there might be two packages being installed

johanneskastl commented:
> I will be looking at this over this weekend to give you full feedback, but in the meantime could you please let me know on which systems have you tested this PR? (I mean OS versions like 'opensuse 15.1', 'SLES 15.1', etc.).
> I'm asking mainly because in metadata section you have noted that this will work on 'all' opensuse versions which I'm worried what 'all' in this context means (all currently supported versions/all existing versions?)

Hi Ondrej,
I have tested using SLES15 SP1, SLES12 SP4 and openSUSE Leap 15.1 as servers, and have found some minor things that I added to this PR. Mainly the packages are different on SLES12.

I have successfully configured a iscsi server using the current state of the PR, and could connect an iniator to this server successfully.

OndrejHome commented on meta/main.yml:
```diff
@@ -9,6 +9,13 @@ galaxy_info:
+    - 15
+  - name: opensuse
+    versions:
+    - all
```
Lets be specific here that this was tested on OpenSuse 15. OpenSUSE 15.x is currently the only supported release. older are in the EOL (end-of-life). Is it OK to change this to `15` only instead of `all`?

johanneskastl replied:
Sure.

johanneskastl replied:
Fixed.

OndrejHome commented on tasks/main.yml:
```diff
@@ -1,12 +1,22 @@
+  when: '(ansible_os_family == "Suse" or ansible_os_family == "openSUSE Leap")'
+
+- include_vars: vars_SUSE_SLES12.yml
+  when: 'ansible_os_family == "Suse" and ansible_distribution_major_version|int < 15'
```
As this will apply only to version `12`, could this be just `== 12` ?

johanneskastl replied:
In theory someone could try to use this with SLES11. SLES11 SP4 is still supported by SUSE.

Not sure if it will work there...

OndrejHome replied:
My practical question here if you have tried/tested it in SLES 11?
This version list doesn't prevent anyone from using it only on specific versions but it gives at least idea on where it should work. My approach to this is to keep the list specific to versions that someone has tried the role on and it worked as intended.
**[?]** Would you be able to test it on SLES 11 so that can be added to that list or should we just leave it out?

OndrejHome commented on tasks/vars_SUSE_SLES12.yml:
```diff
@@ -0,0 +1,6 @@
+# SLES12 has a different package name
+targetcli_package:
+  - targetcli-fb
+  - python-dbus-python
```
addition of `python-dbus-python` is a nice catch. I had my testing already preinstalled with this package, but testing to remove I could see that `targetcli` broke. I would expect this to be handled by packaging system (at least in OSes I use) but here it looks it needs to be specified explicitly. Just wanted to say 'nice job finding this!'.

OndrejHome replied:
this doesn't need any changes, it is here as 'thanks for noticing this'. Feel free to 'resolve' this after reading.

OndrejHome reviewed (changes requested):
Thank you for your patience.
In addition to 2 comments I have following requests:
- (medium): move the `vars_*` files from `tasks/` to `vars/` to keep the 'role' convention (it will actually work without any code change)
- (low): add to README information about need to use 'python3' with SLES/openSUSE 15.x (as you can see the explicit `ansible_python_interpreter=/usr/bin/python3` in hosts file below)

I have tested this PR with following 'hosts' file which indicates also versions I have tested on.
~~~
[servers]
# openSUSE 15.1
192.168.22.53 ansible_python_interpreter=/usr/bin/python3
# SLES 15.1
192.168.22.54 ansible_python_interpreter=/usr/bin/python3
# SLES 12.5
192.168.22.56
~~~

Nothing significant blocking this from merge after incorporating the requests made here. Good job preparing and testing this out!

johanneskastl commented:
> Thank you for your patience.
> In addition to 2 comments I have following requests:
> 
>     * (medium): move the `vars_*` files from `tasks/` to `vars/` to keep the 'role' convention (it will actually work without any code change)

I thought files in `vars/` are always being read?

johanneskastl commented:
> Thank you for your patience.
> In addition to 2 comments I have following requests:
> 
>     * (medium): move the `vars_*` files from `tasks/` to `vars/` to keep the 'role' convention (it will actually work without any code change)
> 
>     * (low): add to README information about need to use 'python3' with SLES/openSUSE 15.x (as you can see the explicit `ansible_python_interpreter=/usr/bin/python3` in hosts file below)

Sure, although this is currently a non-fatal error. And as python2 is end of life, I guess this will hit all other distros sooner or later... :-)

Commit:
ADD note for SLES 15.x / openSUSE 15.x python3 to README

johanneskastl commented:
> > Thank you for your patience.
> > In addition to 2 comments I have following requests:
> > ```
> > * (medium): move the `vars_*` files from `tasks/` to `vars/` to keep the 'role' convention (it will actually work without any code change)
> > ```
> 
> I thought files in `vars/` are always being read?

> Basically, anything that goes into “role defaults” (the defaults folder inside the role) is the most malleable and easily overridden. Anything in the vars directory of the role overrides previous versions of that variable in namespace.
https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#playbooks-variables

OndrejHome commented:
> > I thought files in `vars/` are always being read?
> 
> > Basically, anything that goes into “role defaults” (the defaults folder inside the role) is the most malleable and easily overridden. Anything in the vars directory of the role overrides previous versions of that variable in namespace.
> > https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#playbooks-variables

To be honest I had to doeble-check this for myself and result from practical testing is that `role/vars/*` is not loaded automatically. Only `role/vars/main.yml` is loaded and rest needs to be included manually. From the link you have posted it would be

~~~
Here is the order of precedence from least to greatest (the last listed variables winning prioritization):
...
15. role vars (defined in role/vars/main.yml)
...
~~~

https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable

My point here is that this proposed arrangement separates 'task playbooks' from 'vars playbooks/files' into different directories.

Here is a small test for demonstration:

1. including `rrr.yml` via `include_vars` from `vars/` directory.
~~~
$ tree roles/
roles/
└── aaa
    ├── tasks
    │   └── main.yml
    └── vars
        └── rrr.yml

3 directories, 2 files
$ head roles/aaa/*/*
==> roles/aaa/tasks/main.yml <==
---
- include_vars: rrr.yml
- name: test
  debug: var=bbb

==> roles/aaa/vars/rrr.yml <==
---
bbb: 'bbb'
$ ansible-playbook a.yml
PLAY [localhost] ************************************************************

TASK [Gathering Facts] ******************************************************
ok: [localhost]

TASK [aaa : include_vars] ***************************************************
ok: [localhost]

TASK [aaa : test] ***********************************************************
ok: [localhost] => {
    "bbb": "bbb"
}

PLAY RECAP ******************************************************************
localhost: ok=3    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
~~~

2. Not including it and just leaving `rrr.yml` in `vars/` directory. As the variable seems to be not defined I assume that not all files from `vars/` are read by ansible in that case.

~~~
$ head roles/aaa/*/*
==> roles/aaa/tasks/main.yml <==
---
#- include_vars: rrr.yml
- name: test
  debug: var=bbb

==> roles/aaa/vars/rrr.yml <==
---
bbb: 'bbb'
$ ansible-playbook a.yml
PLAY [localhost] ************************************************************

TASK [Gathering Facts] ******************************************************
ok: [localhost]

TASK [aaa : test] ***********************************************************
ok: [localhost] => {
    "bbb": "VARIABLE IS NOT DEFINED!"
}

PLAY RECAP ******************************************************************
localhost : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
~~~

Ansible version for completeness:
~~~
# ansible --version
ansible 2.9.5
  config file = /home/ondrej/.ansible.cfg
  configured module search path = ['/home/ondrej/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib64/python3.6/site-packages/ansible
  executable location = /usr/lib/python-exec/python3.6/ansible
  python version = 3.6.10 (default, Feb 27 2020, 09:44:26) [GCC 9.2.0]
~~~

OndrejHome reviewed (changes requested):
At this point 2 requests from my side:
- (medium) move of `vars_*` files to `vars` directory as explained in earlier comment.
- (low) deciding if this shlould be tested on SLES 11 or not. In my opinion it would be easier to just not test it, but I'm leaving decision up to you to decide if it is worth testing it.

(medium) - needs change in code
(low) - needs only answer (or change in code if 11 should be supported)

Rest of the changes to this point looks good to me, thank you for your continuous effort on this.

Commit:
only use vars_SUSE_SLES12.yml on SLES12

Commit:
use vars_SUSE_SLES11.yml on SLES11

Commit:
MOVE tasks/vars_SUSE_openSUSE.yml to vars/

Commit:
MOVE tasks/vars_RedHat_CentOS.yml to vars/

Commit:
MOVE tasks/vars_SUSE_SLES12.yml to vars/

Commit:
MOVE tasks/vars_SUSE_SLES11.yml to vars/

johanneskastl commented:
> At this point 2 requests from my side:
> 
>     * (medium) move of `vars_*` files to `vars` directory as explained in earlier comment.

Done. Thanks for testing, up until now I read the docs to mean "all files in vars/ are being read".

>     * (low) deciding if this shlould be tested on SLES 11 or not. In my opinion it would be easier to just not test it, but I'm leaving decision up to you to decide if it is worth testing it.

I try to test this tomorrow, and either state that it is working or drop the support.

> Rest of the changes to this point looks good to me, thank you for your continuous effort on this.

You're welcome, I am glad I did not have to write that thing myself from scratch.

johanneskastl commented:
Tested successfully on:

- openSUSE Leap 15.1
- SLES15 SP1
- SLES12 SP4

(using the same versions as clients, but that should not matter)

One thing to note: SLES12 still uses SuSEfirewall2.service, that is not easily managed by ansible. Should we maybe detect that it is running an fail?

johanneskastl commented:
Regarding the test on SLES11, unfortunately I get a python error:
```
   "module_stdout": "\r\nTraceback (most recent call last):\r\n  File \"/home/kastl/.ansible/tmp/ansible-tmp-1585343917.746045-254052704448959/AnsiballZ_targetcli_iscsi_lun.py\", line 102, in <module>\r\n    _ansiballz_main()\r\n  File \"/home/kastl/.ansible/tmp/ansible-tmp-1585343917.746045-254052704448959/AnsiballZ_targetcli_iscsi_lun.py\", line 94, in _ansiballz_main\r\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\r\n  File \"/home/kastl/.ansible/tmp/ansible-tmp-1585343917.746045-254052704448959/AnsiballZ_targetcli_iscsi_lun.py\", line 40, in invoke_module\r\n    runpy.run_module(mod_name='ansible.modules.targetcli_iscsi_lun', init_globals=None, run_name='__main__', alter_sys=True)\r\n  File \"/usr/lib64/python2.6/runpy.py\", line 136, in run_module\r\n    fname, loader, pkg_name)\r\n  File \"/usr/lib64/python2.6/runpy.py\", line 54, in _run_module_code\r\n    mod_loader, pkg_name)\r\n  File \"/usr/lib64/python2.6/runpy.py\", line 34, in _run_code\r\n    exec code in run_globals\r\n  File \"/tmp/ansible_targetcli_iscsi_lun_payload_nVDeMP/ansible_targetcli_iscsi_lun_payload.zip/ansible/modules/targetcli_iscsi_lun.py\", line 154, in <module>\r\n  File \"/tmp/ansible_targetcli_iscsi_lun_payload_nVDeMP/ansible_targetcli_iscsi_lun_payload.zip/ansible/modules/targetcli_iscsi_lun.py\", line 113, in main\r\nIndexError: list index out of range\r\n",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
```

I guess the targetcli package is maybe just too old. Unless you have an idea I would remove the SLES11 things again, and only support SLES12/SLES15 or Leap 15.x.

Commit:
ADD correct service name to vars/vars_SUSE_SLES11.yml

OndrejHome commented on README.md:
```diff
@@ -78,6 +78,16 @@ You can even later instruct the nodes to connect to target created here.
+Note for SUSE Linux Enterprise Server 15.x / openSUSE Leap 15.x
+-------
+
+Please note that SLES 15.x / openSUSE Leap 15.x only ship with python3 installed by default. Therefore the python3 package `python3-targetcli-fb` is being installed on those platforms.
```
(low/nitpick) I would remove the sentence 'Therefore the python3 package `python3-targetcli-fb` is being installed on those platforms.' from here. The important part is that python3 is used and that is why the change needs to be done. Shorter documentation is usually easier to read to users.

johanneskastl replied:
Done.

OndrejHome reviewed (changes requested):
Hi @johanneskastl,
Thank you for the changes!

- [ ] **In regards to SLES 11:**
I would suspect/guess either python-2.6 issue with arrays or different output from targetcli in that version that breaks parsing of data. My opinion on this is that I would rather NOT support it at this moment. However as you have done some testing on SLES 11 I would at least propose adding information to README stating something like:
*This role was tested to not work on SLES 11 and if anyone is interested the PR for adding support is welcomed - for more details refer to PR #4.*

- [ ] **SLES12 still uses SuSEfirewall2.service**
> Should we maybe detect that it is running an fail?

We can either:
A) ignore this situation and mention it only in README
B) fail, but allow the user to skip the fail if they know what they are doing or when they have configured the firewall manually. For example we can introduce new variable `ignore_sles12_firewall_configuration` defaulting to `False`. Later when playbook detects that we are on SLES 12, SuSEfirewall2.service is running and `ignore_sles12_firewall_configuration` is False we can print error saying something like:
 *This role cannot configure firewall on SLES12, you can either configure port 3260/tcp manually for this service and then set `ignore_sles12_firewall_configuration` variable to `True` for this role or stop the SUSE2firewall.service*

My goal here is to allow the user to run role without need to modify the role itself (modifying how the role is called is OK as that happens in users playbook and not roles files)

- [ ] I have added also one tiny idea for README change with low importance - comment in code.

Have a good weekend!

Commit:
README.md: remove sentence

Commit:
README.md: add note on SLES11 not working

Commit:
fail if SuSEfirewall2 is being detected. This can be overridden by setting the variable ignore_sles12_firewall_configuration to True

Commit:
README.md: note on SLES12 with SuSEfirewall2

Commit:
tasks/main.yml: make SuSEfirewall2.service detection not fail on systems not having this service

johanneskastl commented:
Hi Ondrej,

I have
- added a note on SLES11 to the README
- added the necessary configuration so this role fails on hosts running SuSEfirewall2, as discussed
- added a note on SLES12 with SuSEfirewall2 to the README
- remove that one sentence from README

I found that the service_facts task does not always reflect changes in the SuSEfirewall2 service, i.e. after stopping it it is still shown as started. I try to gather more information and create a bug report for that, so nothing to be worked around here.

OndrejHome commented:
Hi @johanneskastl ,

Thank you also for changes here. Same as in the iscsiadm repository this looks good to me by just looking at code. however I would like to test it myself first before merging it. I will get to this over this week. Thank you for your patience and contributions.

OndrejHome commented:
Hi @johanneskastl, I have tested the changes on following systems without any major issues:
SLES12, SLES15, OpenSUSE 15, CentOS 7.
I have added one small change to apply firewalld configuration immediately as same change I have done in past in my other roles. I have also added `name` modules to includes that didn't have them (I'm using `ansible-lint` to check for such things). Also I have split the `fail` text and condition into multiple lines for better readability and to make `ansible-lint` happy.
You can find the final changed role under branch `5.0.0` that was also pushed to ansible galaxy now.

Thank you for your time and contribution to this role!
